File: modules/perf-memory/android/src/main/java/expo/modules/perfmemory/PerfMemoryModule.kt
package expo.modules.perfmemory

import android.os.Debug
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class PerfMemoryModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("PerfMemory")

    // Total PSS (proportional set size) of the process, in bytes. Closest
    // Android analogue to iOS phys_footprint for app-level memory tracking.
    Function("memoryFootprint") {
      val info = Debug.MemoryInfo()
      Debug.getMemoryInfo(info)
      info.totalPss.toLong() * 1024.0
    }

    // Native heap bytes currently allocated. Drops on free(), so it reveals
    // deterministic release (analogue of iOS malloc size_in_use).
    Function("mallocInUse") {
      Debug.getNativeHeapAllocatedSize().toDouble()
    }
  }
}

File: modules/perf-memory/ios/PerfMemoryModule.swift
import ExpoModulesCore

public class PerfMemoryModule: Module {
  public func definition() -> ModuleDefinition {
    Name("PerfMemory")

    // Physical memory footprint in bytes. This is the same metric Xcode's
    // memory gauge reports (task_vm_info.phys_footprint) and the closest
    // proxy to what iOS uses for jetsam/memory-pressure decisions.
    Function("memoryFootprint") { () -> Double in
      var info = task_vm_info_data_t()
      var count = mach_msg_type_number_t(
        MemoryLayout<task_vm_info_data_t>.size / MemoryLayout<integer_t>.size)
      let result = withUnsafeMutablePointer(to: &info) {
        $0.withMemoryRebound(to: integer_t.self, capacity: Int(count)) {
          task_info(mach_task_self_, task_flavor_t(TASK_VM_INFO), $0, &count)
        }
      }
      guard result == KERN_SUCCESS else { return -1 }
      return Double(info.phys_footprint)
    }

    // Live heap currently allocated by malloc across all zones, in bytes.
    // Unlike phys_footprint, this DROPS immediately when memory is free()'d —
    // the same signal Xcode Instruments "Allocations" shows. Best metric for
    // observing deterministic (eager) release. Note: GPU/Metal texture memory
    // is allocated outside malloc, so it isn't counted here.
    Function("mallocInUse") { () -> Double in
      var stats = malloc_statistics_t()
      malloc_zone_statistics(nil, &stats)
      return Double(stats.size_in_use)
    }
  }
}
